### DISCUSSION TITLE
liburing integration?

### DISCUSSION THREAD

**Discussion by ghost** [uNetworking/uWebSockets#1522] [Ideas]
Hi,
Is there a plan to introduce liburing in the code?
It could apparently make the runtime faster, but that's just opinion based out of readings.

**Comment by uNetworkingAB** [comment]
It's not a mature tech yet, people jump on it because of hype but there's still many years before any of that becomes usable with benefit. So far I still see epoll win in benchmarks on my computers. Maybe when I get new ones I will revisit but there's also a huge drawback of io-uring since it is proactor

**Reply by uNetworkingAB** [reply]
There's also little benefit compared to sendmmsg/recvmmsg already handling batches

**Reply by ghost** [reply]
Thanks, yes rhe current performance is already really good.
Have a nice day

**Reply by cirospaciari** [reply]
I think libuv just started io_uring integration (not release yet) maybe in next release

**Comment by uNetworkingAB** [comment]
I need to buy an AMD based computer for this, I have old Intel machines that see no improvement with io_uring at all

**Comment by ghost** [comment]
No amd here as well, I run an i7 8750H with the latest kernel, I will try to give it a look. But I am overwhelmed by work at the moment.
If you have a benchmark ready to run, I will happily run it on my machine

**Comment by cirospaciari** [comment]
I have an I7 7700HQ and one with Ryzen 5500U, if you have some benchmarks that I can run just send me, I'm on Debian bookworm kernel 6.0, and I have some people with Zen 3 5950X that could run the tests for me

**Comment by ghost** [comment]
As an addendum: [test from Alibabacloud](https://www.alibabacloud.com/blog/io-uring-vs--epoll-which-is-better-in-network-programming_599544)

**Comment by uNetworkingAB** [comment]
Yes, it's a low prio thing that will happen when the tech matures. Like said, I have the M1 Air, a tower PC, a laptop, a raspberry pi 4 and I have tested io_uring with every Linux kernel since the inception of io_uring and epoll always wins or loses insignificantly on these machines. I know some AMD machines are faster with io_uring, but it varies.

It has gotten better for io_uring with time (it used to be WAY slower than epoll!), but so far it's inconclusive. As soon as io_uring is undoubtly faster, always, a backend will be written in uSockets (this does not require any change to uWebSockets to take advantage of).

But right now I am just waiting for the tech to get optimized / mature more. Trust me, when it is undoubtly faster I will add it as a high prio item.

**Reply by ghost** [reply]
I literally have trust in you, that article was just to back your claims up and for those waiting for it, to temper the urge. Thanks for your wonderful work.

**Reply by uNetworkingAB** [reply]
I just ran this again https://github.com/frevib/io_uring-echo-server and I get about the same perf. with uSockets or even worse, at the same CPU usage so the benefits are insignificant or entirely missing. Again there are benefits to reactor since it consumed way less memory and if we cannot see significant improvements there is no reason to trade this benefit for nothing.

**Reply by uNetworkingAB** [reply]
I mean the ultimate challenge / proof is pretty simple:

Show code of a server and client written to use io_uring that ping/pongs short TCP messages faster than uSockets (http_load_test and http_server) does it, and I will validate that code here. If any such code exists, I will implement io_uring. Else, I will not.
